### PR TITLE
Refactor web UI file transfers to use sftp package

### LINKED ISSUE #23595: WebUI should use SFTP instead of SCP
WebUI still uses SCP https://github.com/gravitational/teleport/blob/d2193dbf794895254e1b62d29bb091eebc55e564/lib/web/files.go#L118-L146
We should start using SFTP instead.

### PR DESCRIPTION
**Pull request by capnspacehook**
The sftp package is where modern file transfer logic lives and is being maintained. Make the web UI use this package to unify how we transfer files.

Closes https://github.com/gravitational/teleport/issues/23595.

### CODE REVIEW

**Comment by zmb3** [comment]
Please make sure file transfers still work when per session MFA is enabled. 

Cc @avatus

File: lib/sshutils/sftp/http.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/trace"
+)
+
+type ContextKey string
```
Comment (zmb3) [inline lib/sshutils/sftp/http.go:35]: While you're here, can you unexport this type?

File: lib/sshutils/sftp/parse_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			dest:    Destination{Login: "complex@example.com", Host: utils.NetAddr{Addr: "remote.com", AddrNetwork: "tcp"}, Path: "/anything.txt"},
+		},
+		{
+			comment: "implicit user's home directory",
+			in:      "root@remote.host:",
+			dest:    Destination{Login: "root", Host: utils.NetAddr{Addr: "remote.host", AddrNetwork: "tcp"}, Path: "."},
+		},
+	}
+	for _, tt := range testCases {
+		tt := tt
```
Comment (zmb3) [inline lib/sshutils/sftp/parse_test.go:74]: This isn't necessary here.

File: lib/sshutils/sftp/sftp.go [hunk 3]
```diff
@@ -138,6 +143,87 @@ func CreateDownloadConfig(src, dst string, opts Options) (*Config, error) {
 	return c, nil
 }
 
+// HTTPTransferRequest describes file transfer request over HTTP.
+type HTTPTransferRequest struct {
+	// Dst is the source file name
```
Comment (zmb3) [inline lib/sshutils/sftp/sftp.go:148]: Copy pasta error

File: lib/sshutils/sftp/sftp.go [hunk 4]
```diff
@@ -138,6 +143,87 @@ func CreateDownloadConfig(src, dst string, opts Options) (*Config, error) {
...
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// CreateHTTPDownloadConfig returns a Config ready to download a file
+// over HTTP.
+func CreateHTTPDownloadConfig(req HTTPTransferRequest) (*Config, error) {
+	if req.Src == "" {
```
Comment (zmb3) [inline lib/sshutils/sftp/sftp.go:200]: This code is duplicated above. How about adding a checkandsetdefaults?

File: lib/sshutils/sftp/sftp.go [hunk 5]
```diff
@@ -156,8 +242,17 @@ func (c *Config) setDefaults() {
 }
 
 // TransferFiles transfers files from the configured source paths to the
-// configured destination path over SFTP
+// configured destination path over SFTP or HTTP depending on the Config.
 func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	// if this is a transfer over HTTP, no SFTP client is needed
+	if c.isHttpTransfer {
```
Comment (zmb3) [inline lib/sshutils/sftp/sftp.go:248]:
```suggestion
	if c.isHTTPTransfer {
```

File: lib/sshutils/sftp/sftp_test.go [hunk 6]
```diff
@@ -583,6 +587,83 @@ func TestCopyingSymlinkedFile(t *testing.T) {
 	checkTransfer(t, false, dstPath, linkPath)
 }
 
+func TestHTTPUpload(t *testing.T) {
+	tempDir := t.TempDir()
+	src := "source"
+	dst := filepath.Join(tempDir, "destination")
+
+	createFile(t, tempDir, src)
+	src = filepath.Join(tempDir, src)
```
Comment (zmb3) [inline lib/sshutils/sftp/sftp_test.go:596]: It's a little weird for src to start out with one value and then be overwritten. It doesn't seem necessary here.

File: lib/web/files.go [hunk 7]
```diff
@@ -105,102 +105,59 @@ func (h *Handler) transferFile(w http.ResponseWriter, r *http.Request, p httprou
...
-
-	err = tc.ExecuteSCP(ctx, req.serverID, cmd)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	return nil
+type fileTransfer struct {
+	// ctx is a web session context for the currently logged in user.
+	ctx           *SessionContext
```
Comment (zmb3) [inline lib/web/files.go:158]: Let's name this sctx. We should reserve ctx for context.Context values

**Comment by avatus** [comment]
> Please make sure file transfers still work when per session MFA is enabled.
> 
> Cc @avatus

Tested locally and the current iteration has per-session-mfa still working. I also love that this doesn't change anything on the frontend so, 👍 from me

**Comment by avatus** [comment]
A couple notes from local testing.
1. file transfers are way faster now, super sweet!!
2. I recently merged a PR to let this through [if the request includes some env vars related to a moderated session](https://github.com/gravitational/teleport/pull/23875/files#). You have included the env vars in this PR which is great thank you!
3. Downloading file transfers aren't blocked on the web after this change. `tsh tcp` was fixed [here](https://github.com/gravitational/teleport/pull/17679) by adding in the `CheckSFTPAllowed` but I think we are skipping this now on the web with the `c.IsHTTPResponse` block. I think we'll have to do some check similar to my linked PR above where we prevent downloads from a user that requires moderation unless those envvars exist/are valid. Let me know what you think!

File: lib/sshutils/sftp/parse_test.go (hunk 2 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/parse_test.go:74]: I agree, I copied this code from the `scp` package and didn't look too closely at it, but I'll fix it

File: lib/sshutils/sftp/http.go (hunk 1 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/http.go]: Done

File: lib/sshutils/sftp/parse_test.go (hunk 2 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/parse_test.go]: Done

File: lib/sshutils/sftp/sftp.go (hunk 4 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/sftp.go]: Good idea, done

File: lib/sshutils/sftp/sftp_test.go (hunk 6 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/sftp_test.go]: Now that you mention it that's true. I changed `createFile` and `createDir` to take a single path so I could clean instances like this up.

File: lib/web/files.go (hunk 7 above)
Comment (capnspacehook) [reply lib/web/files.go]: Done

**Comment by capnspacehook** [comment]
@avatus thanks for manually testing! I think I fixed the file transfer check, can you take a look and make sure everything looks good?

**Review by avatus — APPROVED** [review]
Looks and works great!

File: lib/web/files.go [hunk 8]
```diff
@@ -105,102 +105,59 @@ func (h *Handler) transferFile(w http.ResponseWriter, r *http.Request, p httprou
...
-}
-
-func (f *fileTransfer) download(req fileTransferRequest, httpReq *http.Request, w http.ResponseWriter) error {
-	ctx := httpReq.Context()
-	cmd, err := scp.CreateHTTPDownload(scp.HTTPTransferRequest{
-		RemoteLocation: req.remoteLocation,
-		HTTPResponse:   w,
-		User:           f.ctx.GetUser(),
-	})
+	tc, err := ft.createClient(req, r)
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/web/files.go:127]: I'll open an issue to track that :+1:

**Comment by capnspacehook** [comment]
Friendly ping @strideynet

File: lib/srv/ctx.go [hunk 9]
```diff
@@ -751,7 +750,20 @@ func (c *ServerContext) CheckSFTPAllowed() error {
...
 		return trace.Wrap(err)
 	}
-	if !canStart {
+	// canStart will be true for non-moderated sessions. If canStart is false, check to
+	// see if the request has been approved through a moderated session next.
+	if canStart {
+		return nil
+	}
+	if registry == nil {
+		return errCannotStartUnattendedSession
```
Comment (strideynet) [inline lib/srv/ctx.go:759]: Should these errors be wrapped so we get a trace to within this function ?

**Review by strideynet — APPROVED** [review]

File: lib/srv/ctx.go (hunk 9 above)
Comment (capnspacehook) [reply lib/srv/ctx.go:759]: I actually was thinking about that and forgot to do something about it... These should probably be normal errors constructed from `errors.New` and wrapped when returned, yeah.

**Pull request merged**